plugin/pkg/rqdata/rqdata.go

```diff
@@ -0,0 +1,220 @@
+package rqdata
+
+import (
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+type requestFunc func(state request.Request) string
+
+// Mapping define the mapping between 'name' of data and the way to extract that data from the Request
+// it also defines what will be the empty value returned if the data behind the name is empty.
+// it is pretty static, and you should need to instantiate only once
+type Mapping struct {
+	replacements map[string]requestFunc
+	emptyValue   string
+}
+
+// Extractor implements a Value(name) (value, valid) function
+// which allow to extract data from an existing DNS Request(or state)
+type Extractor struct {
+	state     request.Request
+	requester *Mapping
+}
+
+//NewExtractor return a new Extractor based on the Mapping and the Request provided
+func NewExtractor(r request.Request, m *Mapping) *Extractor {
+	return &Extractor{r, m}
+}
+
+// NewMapping build the mapping name -> func to extract data from the Request
+func NewMapping(emptyValue string) *Mapping {
+	replacements := map[string]requestFunc{
+		"type": func(state request.Request) string {
```

Comment:
Why does all this stuff require a expansive setup of getting these values? you can just return state.Type() with the functions and without the map

Reply:
The idea is to build the Mapping only once, when creating the plugin and reuse the same mapping with every DNS query coming in.
=> If I use direct function, I end-up creating xomething similar to `replacer` that created the map on each DNS request.

That is the performance improvement proposed here : https://github.com/coredns/coredns/issues/2293